ipaserver/plugins/baseuser.py

```diff
@@ -728,3 +739,138 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
+        DNParam(
+            'subject?',
+            cli_name='subject',
+            label=_('Subject'),
+            doc=_('Subject of the certificate'),
+            flags=['virtual_attribute', 'no_create', 'no_update']
+        ),
+        Bytes(
+            'usercertificate*', validate_certificate,
+            cli_name='certificate',
+            label=_('Certificate'),
+            doc=_('Base-64 encoded user certificate'),
+        ),
+    )
+
+    @staticmethod
+    def _build_mapdata(subject, issuer):
+        return u'X509:{issuer}{subject}'.format(
+            issuer='<I>{}'.format(issuer.x500_text()) if issuer else '',
+            subject='<S>{}'.format(subject.x500_text()) if subject else '',
+            )
+
+    @staticmethod
+    def _convert_options_to_certmap(certmapdata, options):
+        """
+        Converts options to ipacertmapdata
+
+        When --subject --issuer or --usercertificate options are used,
+        the value for ipacertmapdata is built from extracting subject and
+        issuer,
+        converting their values to X500 ordering and using the format
+        X509:<I>issuer<S>subject
+        For instance:
+        X509:<I>O=DOMAIN,CN=Certificate Authority<S>O=DOMAIN,CN=user
+        A list of values can be returned if --usercertificate is used multiple
+        times, or in conjunction with --subject --issuer.
+        """
+        data = []
+        data.extend(certmapdata)
+
+        if 'issuer' in options or 'subject' in options:
+            issuer = options.get('issuer')
+            subject = options.get('subject')
+            data.append(ModCertMapData._build_mapdata(subject, issuer))
+
+        for dercert in options.get('usercertificate', ()):
+            cert = x509.load_certificate(dercert, x509.DER)
+            issuer = DN(cert.issuer)
+            subject = DN(cert.subject)
+            data.append(ModCertMapData._build_mapdata(subject, issuer))
+
+        return data
+
+    def get_args(self):
+        # ipacertmapdata is not mandatory as it can be built
+        # from the values subject+issuer or from reading usercertificate
+        for arg in super(ModCertMapData, self).get_args():
+            if arg.name == 'ipacertmapdata':
+                yield arg.clone(required=False, alwaysask=False)
+            else:
+                yield arg.clone()
+
+    def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys,
+                     **options):
+        # The 3 valid calls are
+        # ipa user-add-certmapdata LOGIN --subject xx --issuer yy
+        # ipa user-add-certmapdata LOGIN [DATA] --certificate xx
+        # ipa user-add-certmapdata LOGIN DATA
+        # Check that at least one of the 3 formats is used
+
+        # If only LOGIN is supplied, then we need either subject or issuer or
+        # certificate
+        if len(keys) < 2:
+            valid_options = ('issuer', 'subject', 'usercertificate')
+            if all(key not in options for key in valid_options):
```

Comment:
I'm pretty sure just checking for presence of the options is not good enough, you should also check that none of the options are `None`. IMO the easiest way to achieve this would be to add this at the beginning of `pre_callback`:
```python
        try:
            certmapdatas = args[1] or []
        except IndexError:
            certmapdatas = []
        issuer = options.get('issuer')
        subject = options.get('subject')
        certificates = options.get('usercertificate') or []
```
and work with these variables further in the code.